zerver/tests/test_messages.py

```diff
@@ -3528,6 +3522,24 @@ def test_claim_attachment(self) -> None:
             attachment = Attachment.objects.get(path_id=path_id)
             self.assertTrue(attachment.is_claimed())
 
+class LinkDetectionTest(ZulipTestCase):
+    def test_finds_all_links(self) -> None:
+        msg_ids = []
+        msg_contents = ["foo.org", "[bar](baz.gov)", "http://quux.ca"]
+        for msg_content in msg_contents:
+            msg_ids.append(self.send_stream_message(self.example_email('hamlet'),
+                                                    'Denmark', content=msg_content))
+        msgs = [Message.objects.get(id=id) for id in msg_ids]
+        self.assertTrue(all([msg.has_link for msg in msgs]))
+
+    def finds_link_after_edit(self) -> None:
+        msg_id = self.send_stream_message(self.example_email('hamlet'),
+                                          'Denmark', content='a')
+        msg = Message.objects.get(id=msg_id)
+        self.assertFalse(msg.has_link)
+        do_update_message(self.example_user('hamlet'), msg, content="a http://foo.com")
+        self.assertTrue(msg.has_link)
```

Comment:
Similarly here, a negative test would be useful. A brief reading of the changes suggests we don't have any way to set it to False if an edit removes all links in the message.

Maybe we could init has_link to False before starting to render the message in bugdown.

Reply:
Did this in the update message logic actually - thanks for pointing this out!

Reply:
Never mind about the earlier comment - my way didn't work at all but I had made a mistake on the test. Doing it as you suggested!